### PR TITLE
refactor(*): dev env setup, code cleanup

### PR DESCRIPTION
**Pull request by zaida04**
> ### ⚠️ Don't merge yet! This PR is awaiting testing and review from @vladfrangu

## What does this PR accomplish?
- [x] ESLint & Prettier (and config)
- [x] Husky Pre-commit hooks (in connection with lint-staged and commitlint)
- [x] General "makeRequest" function that all fetch requests are abstracted to.
- [x] Fully type all outgoing requests made so that the code is type-safe.
- [x] Cleanup of the code (condensing variables, renaming, etc.)
- [ ] ~Handle dAPI ratelimits/retry (?? is this needed? It's an oauth application so can we afford to wait/retry requests?)~

## That's cool, why?
image

### CODE REVIEW

**Comment by jb3** [comment]
API ratelimits are likely not necessary to handle. We won't hit them.

File: src/bindings.d.ts [hunk 1]
```diff
@@ -1,9 +1,11 @@
+import { Snowflake } from 'discord-api-types'
```
Comment (vladfrangu) [inline src/bindings.d.ts]: Please import from a version or from /globals

File: src/index.ts [hunk 2]
```diff
@@ -1,112 +1,84 @@
 import { getErrorPageHTML, Router } from '8track'
```
Comment (vladfrangu) [inline src/index.ts]: It seems like eslint isn't linting these files, no semis?

File: src/index.ts [hunk 3]
```diff
@@ -1,112 +1,84 @@
...
+		return ctx.end('Missing authorization code', { status: 400 })
+	}
+
+	try {
+		const accessTokenReq = await getAccessToken(code)
+		const accessTokenBody = await accessTokenReq[0]
+		const accessToken = accessTokenBody['access_token']
+		const userGuilds = await getUserGuilds(accessToken)
+
+		if (!userGuilds.some((x) => WHITELISTED_GUILDS.includes(x)))
```
Comment (vladfrangu) [inline src/index.ts]: Faster to reverse this some check (check if either allowed guild is in the users guild

File: src/oauth2.ts [hunk 4]
```diff
@@ -1,96 +1,68 @@
...
-  return await fetch(`${API_BASE}/oauth2/token`, {
-    method: 'POST',
-    headers: {
-      'Content-Type': 'application/x-www-form-urlencoded',
-    },
-    body: data,
-  })
+import { OAuth2Routes, Routes } from 'discord-api-types/v8'
+import { CONSTANTS, makeRequest, UTILS } from './util'
+// eslint-disable-next-line no-duplicate-imports
```
Comment (vladfrangu) [inline src/oauth2.ts]: Don't duplicate imports

File: src/util.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	route: string,
+	options: RequestInit,
+): makeRequestResponseTuple<T> {
+	const req = await fetch(route, options)
+	if (!req.ok) {
+		const parseErrorResponseBody = await req.json().catch(() => ({}))
+		throw new DiscordAPIError(parseErrorResponseBody.message, options.method!, route, req.status)
+	}
+
+	return [req.json() as Promise<T>, req]
```
Comment (vladfrangu) [inline src/util.ts]:
```suggestion
	return [await req.clone().json() as Promise<T>, req]
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: .lintstagedrc.json [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+    "*.ts": "eslint --fix"
```
Comment (vladfrangu) [inline .lintstagedrc.json]: Can you do this like we do in, say, discord-api-types? (the whole --ext and all

File: src/index.ts [hunk 7]
```diff
@@ -1,112 +1,84 @@
...
+	const code = url.searchParams.get('code');
+
+	if (!code) {
+		return ctx.end('Missing authorization code', { status: 400 });
+	}
+
+	try {
+		const accessTokenReq = await getAccessToken(code);
+		const accessTokenBody = await accessTokenReq[0];
+		const accessToken = accessTokenBody['access_token'];
```
Comment (vladfrangu) [inline src/index.ts]: Use dot notation

File: src/index.ts [hunk 8]
```diff
@@ -1,112 +1,84 @@
...
+		if (!WHITELISTED_GUILDS.some((x) => userGuilds.includes(x)))
+			return ctx.end('You are not authorized to join DDD.', { status: 403 });
+
+		const userDetails = await getUserDetails(accessToken);
+		const joinUserToGuildStatus = (await (await joinUserToGuild(accessToken, userDetails.id))[1]).status;
+		const revokeUserTokenResponse = await (await revokeToken(accessToken))[1];
+
+		if (!revokeUserTokenResponse.ok)
+			return ctx.end(
+				`An unexpected error occurred while joining you to the guild. Your authorization token could not be revoked, you can do so in your Discord client under User Settings > Authorized Apps`,
```
Comment (vladfrangu) [inline src/index.ts]: Inaccurate per say, this error is if we couldn't revoke the token, not if we couldn't add you to a guild. Should handle that case too

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by jb3 — APPROVED** [review]

**Pull request merged**